sources.sh

```diff
@@ -39,60 +54,145 @@ for tag in $externalPins; do
+								"parents": { }
+							}
 						}
 					}
-				}
+				{{- end -}}
 			{{- end -}}
 		{{- end -}}
-	{{- end -}}
-' "$@" | jq 3>&1 1>&2 2>&3- -r '
-	# https://github.com/jqlang/jq/issues/2063 - "stderr" cannot functionally output a string correctly until jq 1.7+ (which is very very recent), so we hack around it to get some progress output by using Bash to swap stdout and stderr so we can output our objects to stderr and our progress text to stdout and "fix it in post"
-	# TODO balk / error at multiple arches entries
-	first(.arches | keys_unsorted[]) as $arch
-	| .arches[$arch].tags[0] as $tag
-	| stderr
-	| "\($tag) (\($arch)): \(.sourceId)"
-	# TODO if we could get jq 1.7+ for sure, we can drop this entire "jq" invocation and instead have the reduce loop of the following invocation print status strings directly to "stderr"
-' | jq -n --argjson pins "$externalPinsJson" '
+	' "$@" )
+	if [ -n "$doHeavy" ]; then
+		"${bbCat[@]}" | jq 3>&1 1>&2 2>&3- -r '
+			# https://github.com/jqlang/jq/issues/2063 - "stderr" cannot functionally output a string correctly until jq 1.7+ (which is very very recent), so we hack around it to get some progress output by using Bash to swap stdout and stderr so we can output our objects to stderr and our progress text to stdout and "fix it in post"
+			# TODO balk / error at multiple arches entries
+			first(.arches | keys_unsorted[]) as $arch
+			| .arches[$arch].tags[0] as $tag
+			| stderr
+			| "\($tag) (\($arch)): \(.sourceId)"
+			# TODO if we could get jq 1.7+ for sure, we can drop this entire "jq" invocation and instead have the reduce loop of the following invocation print status strings directly to "stderr"
+		' | jq -n '[ inputs ]'
+	else
+		"${bbCat[@]}" | jq -n '[ inputs ]'
+	fi
+}
+
+# merges heavy-to-calculate data from the second json input (list or map of sources) into the first json input (list of sources)
+# uses "mostlyUniqueBitsSum" as a rough analogue for sourceId to correlate data between the input lists
+#  (sourceId, reproducibleGitChecksum, SOURCE_DATE_EPOCH, froms, lastStageFrom)
+# echo '[{}, {},...] [{extraData},...]' | mergeData
+mergeData() {
+	jq --slurp '
+		def mostlyUniqueBitsSum($archKey):
+			.GitCommit
+			+ .Directory
+			+ .File
+			+ .Builder
+			+ $archKey
+			| @base64
+		;
+		(
+			.[1]
+			| map(
+				. as $source
+				| .arches
+				| keys
+				| map(
+					. as $a
+					| $source.entries[]
+					| {
+						(. | mostlyUniqueBitsSum($a)): { entry: ., source: $source }
+					}
+				)
+			) | add | add
+		) as $cacheFile
+		| .[0]
+		| map(
+			. as $it
+			| ($it.arches | keys_unsorted)[0] as $arch
+			| (
+				# match an item by the unique bits that we have
+				$cacheFile[$it.entries[0] | mostlyUniqueBitsSum($arch)] | select(.source.sourceId) | .entry as $entry | .source | $it * {
+					sourceId,
+					reproducibleGitChecksum,
+					arches: {
+						($arch): {
+							froms: .arches[$arch].froms,
+							lastStageFrom: .arches[$arch].lastStageFrom,
+						},
+					},
```

Comment:
If any of these values in the "cache" are `null`, this will replace them in our file with `null` too? I'm thinking specifically of cases where we might rename or restructure these fields and need to know to ignore the cache but still know to recalculate.  That'll get a "true-up" in the second half of the code below, right?  (Assuming we update the `heavy` calculation to take that into account -- I'll comment separately about that shortly.)

Maybe we can add a comment explaining that? (something like: this might pull in "null" values from the cache if we change the format, but they'll get fixed on the second round of "mergeData" ?  is that accurate?)